doc/designs/quic-design/quic-io-arch.md

```diff
@@ -0,0 +1,485 @@
+QUIC I/O Architecture
+=====================
+
+This document discusses possible implementation options for the I/O architecture
+internal to the libssl QUIC implementation, discusses the underlying design
+constraints driving this decision and introduces the resulting I/O architecture.
+It also identifies potential hazards to existing applications, and identifies
+how those hazards are mitigated.
+
+Objectives
+----------
+
+The OpenSSL QUIC API design is intended to meet the following objectives,
+amongst others:
+
+ - We want to support both blocking and non-blocking semantics
+   for application use of the libssl APIs.
+
+ - In the case of non-blocking applications, it must be possible
+   for an application to do its own polling and make its own event
+   loop.
+
+Requirements
+------------
+
+These requirements are complicated by the fact that traditional use of the
+libssl API allows an application to pass an arbitrary BIO to an SSL object; not
+only that, separate BIOs can be passed for the read and write directions. The
+nature of this BIO can be arbitrary; it could be a socket, or a memory buffer.
+
+Implementation of QUIC will require that the underlying network BIO passed to
+the QUIC implementation be configured to support datagram semantics instead of
+bytestream semantics as has been the case with traditional TLS over TCP.
+
+Implementation of QUIC requires handling of timer events as well as the
+circumstances where a network socket becomes readable or writable. In many cases
+we need to handle these events simultaneously (e.g. wait until a socket becomes
+readable, or a timeout expires, whichever comes first).
+
+Blocking vs. Non-Blocking I/O
```

Comment:
Somewhere there needs to be a discussion on the difference between network level blocking/non-blocking and application level blocking/non-blocking. A lot of the discussion in this document is about how we will provide blocking semantics at the application level and different strategies for doing that, and whether or not we can support network level blocking/non-blocking sockets in doing so. These distinctions are quite confused in the document at the moment.

Reply:
Updated text to clarify this a bit.